force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,14 +260,63 @@ public with sharing class DMLOperations {
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
-
 		// Extract the Account Name from Contact's LastName
-
 		// Get the account based on the name or create one if it doesn't exist
-
 		// Link the Contact to its Account using the Account Id
-
 		// Upsert the Contacts
+
+        List<String> contactLastNames = new List<String>();
+        for (Contact currentContact : contactsToUpsert) {
+            contactLastNames.add(currentContact.LastName);
+        }
+
+        List<Account> existingAccountList = [SELECT Id, Name FROM Account WHERE Name in :contactLastNames];
+        List<Account> newAccountList = new List<Account>();
+        List<Contact> contactsNeedsAccountIdUpdated = new List<Contact>();
+        List<Contact> contactsNeedsNewAccount = new List<Contact>();
+
+        // Create a list of of Contacts that need to have Account Id updated. 
+        if (!existingAccountList.isEmpty()) {
+            for (Account currentAccount1 : existingAccountList) {
+                for (Contact currentContact1 : contactsToUpsert) {
+                    if ((currentContact1.LastName == currentAccount1.Name)) {
+                        currentContact1.AccountId = currentAccount1.Id;
+                        contactsNeedsAccountIdUpdated.add(currentContact1);
+                        break;
+                    }
+                }
+            }
+        }
+
+        // Update Contacts' Account Id with existing accounts. 
+        // Create a list of Contacts that need to a new account created.
+        if (!contactsNeedsAccountIdUpdated.isEmpty()) {
+            upsert contactsNeedsAccountIdUpdated;
+            for (Contact currentContact2 : contactsToUpsert) {
+                if (!contactsNeedsAccountIdUpdated.contains(currentContact2)) {
+                    contactsNeedsNewAccount.add(currentContact2);
+                }
+            }
+        }
+        // Create new Accounts and assign them to Contacts
+        if (!contactsNeedsNewAccount.isEmpty()) {
+            for (Contact currentContact3 : contactsNeedsNewAccount) {
+                Account newAccount = new Account();
+                newAccount.Name = currentContact3.LastName;
+                newAccount.Id = Null;
+                newAccountList.add(newAccount);
+            }
+            insert newAccountList;
+
+            for (Contact currentContact4 : contactsNeedsNewAccount) {
+                for (Account currentAccount4 : newAccountList) {
+                    if (currentAccount4.Name == currentContact4.LastName) {
+                        currentContact4.AccountId = currentAccount4.Id;
+                    }
+                }
+            }
+            upsert contactsNeedsNewAccount;
```

Comment:
Way to work through this in a way that does not do DML inside a loop. I would suggest trying to use Maps to keep track of your existing values, that would avoid needing to do so much looping here. It could be cleaned up a bit and made simpler, but overall very good job in working through this logic.